### DISCUSSION TITLE
AtributeError: __enter__ in python

### DISCUSSION THREAD

**Discussion by LuigiP99** [community/community#30822] [Other Feature Feedback, Questions, & Ideas] [answered]
> Performing a CRUD in python, I created the connection class where is two class methods: getConnection and getCursor, when I created the PersonDAO class and its class method called seleccionar and instantiate the getCursor method which indirectly calls the getConnection method. It throws me the error of AttributeError: \_\_enter\_\_ (details of the error below the code). According to what I read in other forums, I need to define the methods \_\_enter\_\_ and \_\_exit\_\_  in the Conexion class but I don't understand how to do it, any help would be appreciated.

```
from logger_base import log
import psycopg2 as bd
import sys # Modulo que tiene una funcion para terminar por completo nuestro programa

class Conexion:

    _DATABASE = 'test_db'
    _USERNAME = 'postgres'
    _PASSWORD = 'admin'
    _DB_PORT = '5432'
    _HOST = '127.0.0.1'
    _conexion = None
    _cursor = None

    @classmethod
    def obtenerConexion(cls):
        if cls._conexion is None:
            try:
                cls._conexion = bd.connect(host=cls._HOST,
                                           user=cls._USERNAME,
                                           password=cls._PASSWORD,
                                           port=cls._DB_PORT,
                                           database=cls._DATABASE)
                log.debug(f'Conexion exitosa: {cls._conexion}')
                return cls._conexion
            except Exception as e:
                log.error(f'Ocurrio un error al obtener la conexion: {e}')
                sys.exit()

        else:
            return cls._conexion

    @classmethod
    def obtenerCursor(cls):
        if cls._cursor is None:
            try:
                cls._cursor = cls.obtenerConexion().cursor()
                log.debug(f'Se abrio correctamente el cursor: {cls._cursor}')
            except Exception as e:
                log.error(f'Ocurrio una excepcion al obtener el cursor: {e}')
                sys.exit()
        else:
            return cls._cursor


class PersonaDAO:

    '''
    DAO = Data Access Object. 
    Es un patron de diseño para comunicarse con la base de datos. 
    Mejor conocido como operaciones CRUD (Create, read, update & delete)
    '''
    _SELECCIONAR = 'SELECT * FROM persona ORDER BY id_persona'
    _INSERTAR = 'INSERT INTO persona(nombre, apellido, email) VALUES(%s, %s, %s)'
    _ACTUALIZAR = 'UPDATE persona SET nombre=%s, apellido=%s, email=%s WHERE id_persona=%s'
    _ELIMINAR = 'DELETE FROM persona WHERE id_persona=%s'

    @classmethod
    def seleccionar(cls):
        with Conexion.obtenerCursor() as cursor:
            cursor.execute(cls._SELECCIONAR)
            registros = cursor.fetchall()
            personas = []
            for registro in registros:
                persona = Persona(registro[0], registro[1], registro[2], registro[3])
                personas.append(persona)
            return personas


class Persona:
    def __init__(self, id_persona = None, nombre = None, apellido = None, email = None) -> None:
        self._id_persona = id_persona
        self._nombre = nombre
        self._apellido = apellido
        self._email = email

    def __str__(self) -> str:
        return f'''
            Id persona: {self._id_persona}, Nombre: {self._nombre}
            Apellido: {self._apellido}, Email: {self._email}'''

    @property
    def id_persona(self):
        return self._id_persona

    @id_persona.setter
    def id_persona(self, id_persona):
        self._id_persona = id_persona

    @property
    def nombre(self):
        return self._nombre

    @nombre.setter
    def nombre(self, nombre):
        self._nombre = nombre

    @property
    def apellido(self):
        return self._apellido

    @apellido.setter
    def apellido(self, apellido):
        self._apellido = apellido

    @property
    def email(self):
        return self._email

    @email.setter
    def email(self, email):
        self._email = email
```

```
Error: 

10:19:53 PM: DEBUG [conexion.py:24] Conexion exitosa: <connection object at 0x0000020A40C71580; dsn: 'user=postgres password=xxx dbname=test_db host=127.0.0.1 port=5432', closed: 0>
10:19:53 PM: DEBUG [conexion.py:38] Se abrio correctamente el cursor: <cursor object at 0x0000020A42EF5540; closed: 0>
Traceback (most recent call last):
  File "c:\Users\luigi\Desktop\Programacion\CursoPython\Python\CursoUdemy\PostgreSQL\capa_datos_persona\persona_dao.py", line 29, in <module>
    personas = PersonaDAO.seleccionar()
  File "c:\Users\luigi\Desktop\Programacion\CursoPython\Python\CursoUdemy\PostgreSQL\capa_datos_persona\persona_dao.py", line 19, in seleccionar
    with Conexion.obtenerCursor() as cursor:
AttributeError: __enter__
```

**Comment by airtower-luna** [comment, marked as answer]
What happens is that you try to return the cursor returned by `Conexion.obtenerCursor()` as a *context  manager*, but apparently it isn't one:

```
        with Conexion.obtenerCursor() as cursor:
```

A context manager in Python is an object that defines a way to enter a context (for example, acquire a database cursor), and later exit the context (e.g. release the cursor). The documentation describes it in detail in [Context Manager Types](https://docs.python.org/3/library/stdtypes.html#context-manager-types). The nice thing is that a correctly written context manager always tidies up after itself, even if there's an exception in the `with` block.

In many cases you don't need to write your own `__enter__` and `__exit__` methods, though. With the [@contextlib.contextmanager](https://docs.python.org/3/library/stdtypes.html#typecontextmanager) annotation you only need to write a function that creates the resource needed for the context (in your case the cursor), `yield`s it, and then cleans up (e.g. close the cursor). You could easily rewrite your `Conexion.obtenerCursor()` method that way, and get rid of the cursor as a class attribute that way, too (that looks dangerous anyway).

**Reply by LuigiP99** [reply]
Thank you very much for the explanation and for the documentation, I had seen the same thing in other forums and pages, and I know what the error is due to, but I don't know how to work with the context manager yet, I tried to implement it but it gave me an error, the most It's likely that I'm doing it wrong, but I don't know how to implement it, especially when it's at the class level, what I understand is that it must be at the object level, but in this case I am very lost

**Reply by airtower-luna** [reply]
> especially when it's at the class level, what I understand is that it must be at the object level, but in this case I am very lost

You're right, a context manager must be an object. There's no way around that.

You *can* have your class method create the context manager object. `@contextlib.contextmanager` is the easiest way to do that. See the example in the docs I linked above, that should work just as well in a class method as it does in a function.

The other question is whether it *makes sense* to do that in a class function. In programming in general, when you get into a weird place where things don't seem to fit it's often a good idea to take a step back and ask yourself if you're approaching the problem from the right angle. Maybe you are, and you've stepped into weird, uncharted territory, where no-one has built nice APIs yet. But more often then not changing your approach will make things easier.

In a code review I'd reject the attempt to manage a database connection at class level. The class should define the methods necessary to manage it, but as functions on an object of the class. A common approach would be:

* Have a constructor (`__init__`) in the class that initializes the connection.
* Have methods in the class (as regular methods, not class methods) that work with the connection (e.g. return a context manager for running queries).
* Where you actually want to use the database, create an object of the class and work with it. Possibly at a high level so you only need one connection for multiple cursors.

So the question is: What is your goal here? I assume having a class method return a context manager isn't it. :slightly_smiling_face:

**Reply by LuigiP99** [reply]
Thanks buddy, the error is because I have not returned the cursor in the getCursor method and for that detail it does not close by itself, I have realized it by reviewing the code many times and I am surprised that none of the others who ask this error have noticed it haha
